core/include/vecmem/memory/unique_ptr.hpp

```diff
@@ -0,0 +1,462 @@
+#include <type_traits>
+
+#include "vecmem/memory/memory_resource.hpp"
+
+namespace vecmem {
+namespace {
+/**
+ * @brief A deleter class for non-trivial objects.
+ *
+ * This class provides all the necessary functionality to allow unique pointers
+ * to deallocate non-trivial objects, as well as arrays of them.
+ *
+ * Object pointers not only deallocate their memory, but they also cleanly
+ * delete the object by callings its deconstructor.
+ *
+ * The design of this class is somewhat unconventional because it is stateful,
+ * which most deleters are not. However, this is pretty much necessary we
+ * cannot rely on the packed storage of allocation parameters that you see in
+ * standard deleters. In particular, the pointer argument is not strictly
+ * necessary, but it is included to make the class easier to debug.
+ *
+ * @tparam T The type to deallocate.
+ *
+ * @note This deleter class supports non-array types and array types of unknown
+ * bound, but it does not support arrays of known bound.
+ *
+ * @warning This type should never be used with host-inaccessible memory! We
+ * cannot check this at compile-time or at run-time, so it is the
+ * responsibility of the user to ensure that this requirement is respected.
+ */
+template <typename T,
+          std::enable_if_t<!std::is_array_v<T> ||
+                               (std::is_array_v<T> && std::extent_v<T> == 0),
+                           bool> = true>
+struct unique_obj_deleter {
```

Comment:
Does this need to be a template? :thinking: The exact type is not needed by the class. It just needs to know whether the type is an array or not.

I would rather create two different deleter classes. One meant for arrays, and one for single objects. We could make `vecmem::unique_obj_ptr` choose between the two using a traits class. Like:

```c++
template <typename T>
using unique_obj_ptr = std::unique_ptr<T, details::unique_obj_deleter<T>::type>;
```

I'm just a pretty keen on putting as much of the code into compiled sources as possible, and only keeping stuff templated that really needs to be.